apps/crn-server/src/app.ts

```diff
@@ -254,9 +271,33 @@ export const appFactory = (libs: Libs = {}): Express => {
   const pageDataProvider = isContentfulEnabled
     ? pageContentfulDataProvider
     : pageSquidexDataProvider;
+
+  featureFlagDependencySwitch.setDependency(
+    'teams',
+    libs.teamSquidexDataProvider ||
+      new TeamSquidexDataProvider(squidexGraphqlClient, teamRestClient),
+    'IS_CONTENTFUL_ENABLED_V2',
+    false,
+  );
+
+  featureFlagDependencySwitch.setDependency(
+    'teams',
+    libs.teamContentfulDataProvider ||
+      new TeamContentfulDataProvider(
+        contentfulGraphQLClient,
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
above I defined `contentfulRestClient` like this:

```
  let contentfulRestClient: Environment | undefined;

  (async () => {
    contentfulRestClient = await getContentfulRestClient({
      space: contentfulSpaceId,
      accessToken: contentfulManagementAccessToken,
      environment: contentfulEnvId,
    });
  })();
```

so here I was seeing an error that `contentfulRestClient` was being used before assignment. This non-null assertion avoids this error, I know it's not ideal but I could not find a better way to do that.

I haven't done this

```
const contentfulRestClient = await getContentfulRestClient({
    space: contentfulSpaceId,
    accessToken: contentfulManagementAccessToken,
    environment: contentfulEnvId,
  });
```

because then I would need to add `async` in `appFactory` and this seems a little weird to me. Also, this would make it necessary to call appFactory like this

```
  (async () => {
    app = await appFactory({
      ...
    });
  })();
```

where currently is called at the top level

Reply:
I understand what you have done but it's a little risky.

I'm surprised this even works, the DI might actually be resolved before the var is assigned, I guess it's because when you pass this variable in the constructor it's still a reference, not a copy. 

I think we should avoid such risky patterns. Why don't you define this dependency as Promise<Environment> and then just make sure to await it before using inside `TeamContentfulDataProvider`? Yes I know this means having to await each time but I think that's still better.

Alternatively you could look into top level await - it should be supported in our node version, this should mean that we should be able to make the appFactory async and then await it rather than use the pattern you suggested.

@lennym @Riccardo-Brown thoughts?

Reply:
I agree that passing the promise is probably the best approach here. All of the methods on the client are async already, so it's fairly straightforward to replace `const environment = this.restClient;` in the data provider with `const environment = await this.restClient;`.